File: src/CmdShrink.cpp
// CmdShrink.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdShrink.C

#include "CmdShrink.h"
#include <QDebug>
#include <math.h>
#include "Error.h"

static CBuilder<CmdShrink> cbShrink("shrink");

#define SHRINK_DEFAULT 0.2

bool CmdShrink::usage() {
  return error("Usage: shrink [margin_pt] [ratio]\n");
}

bool CmdShrink::parse(Statement const &s) {
  if (s.length()==1)
    return true;
  else if (s.length()==2 && s[1].typ==Token::NUMBER)
    return true;
  else if (s.length()==3 && (s[1].typ==Token::NUMBER || s[1].typ==Token::DASH)
	   && (s[2].typ==Token::NUMBER))
    return true;
  else
    return usage();
}

void CmdShrink::render(Statement const &s, Figure &f, bool) {
  double mrg = pt2iu(SHRINK_DEFAULT);
  if (s.length()>=2 && s[1].typ==Token::NUMBER)
    mrg = pt2iu(s[1].num);
  bool hasRatio = s.length()>=3;
  double ratio = hasRatio ? s[2].num : 1;
  QRectF actual = f.fullBBox();
  QRectF desired = f.extent();

  double dleft = actual.left() - desired.left(); // +ve means it's OK
  double dright = desired.right() - actual.right();
  double dtop =  actual.top() - desired.top();
  double dbottom = desired.bottom() - actual.bottom();

  QPointF x0 = f.xAxis().minp();
  QPointF x1 = f.xAxis().maxp();
  QPointF y0 = f.yAxis().minp();
  QPointF y1 = f.yAxis().maxp();

  double olddx = x1.x()-x0.x();
  double olddy = y1.y()-y0.y();
  
  if (dleft<mrg)
    x0 += QPointF(2*mrg-dleft, 0);
  if (dright<mrg)
    x1 -= QPointF(2*mrg-dright, 0);
  if (dtop<mrg)
    y1 += QPointF(0, 2*mrg-dtop);
  if (dbottom<mrg)
    y0 -= QPointF(0, 2*mrg-dbottom);

  double newdx = x1.x()-x0.x();
  double newdy = y1.y()-y0.y();

  if (newdx*olddx<0 || newdy*olddy<0) {
    f.markFudged();
    //    qDebug() << "Shrink at " << s.label() << " did work";
    return;
  }

  if (hasRatio) {
    double dpx = fabs(x1.x() - x0.x());
    double dpy = fabs(y1.y() - y0.y());
    double xrat = fabs(dpx / (f.xAxis().max() - f.xAxis().min()));
    double yrat = fabs(dpy / (f.yAxis().max() - f.yAxis().min()));
    double myrat = yrat/xrat;
    if (myrat>ratio) {
      // I am too tall
      y1 += QPointF(0, .5*dpy*(1-ratio/myrat));
      y0 -= QPointF(0, .5*dpy*(1-ratio/myrat));
    } else if (myrat<ratio) {
      // I am too wide
      x1 -= QPointF(.5*dpx*(1-myrat/ratio), 0);
      x0 += QPointF(.5*dpx*(1-myrat/ratio), 0);
    }
  }      

  if ((x0 - f.xAxis().minp()).manhattanLength()>mrg ||
      (x1 - f.xAxis().maxp()).manhattanLength()>mrg ||
      (y0 - f.yAxis().minp()).manhattanLength()>mrg ||
      (y1 - f.yAxis().maxp()).manhattanLength()>mrg) {
    f.xAxis().setPlacement(x0, x1);
    f.yAxis().setPlacement(y0, y1);
    f.markFudged();
    //qDebug() << "Shrink at " << s.label() << " did work";
  }
}

File: backend/python/cmdshrink.py
# cmdshrink.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

import math
from typing import TYPE_CHECKING

from PyQt6.QtCore import QPointF, QRectF

from .command import Command
from .token_ import Token
from .factor import pt2iu
from .error import Error

if TYPE_CHECKING:
    from .statement import Statement
    from .figure import Figure

_SHRINK_DEFAULT = 0.2  # pt


@Command.register("shrink")
class CmdShrink(Command):
    """Reduce axis placement to fit content inside the panel extent.

    Syntax:
        shrink
        shrink margin_pt
        shrink margin_pt ratio

    Default margin is 0.2 pt. Default ratio is unspecified (x and y
    are adjusted independently). Does not work for rotated axes, and
    assumes the y-axis runs up.
    """

    def _usage(self) -> bool:
        return self.error("Usage: shrink [margin_pt] [ratio]\n")

    def parse(self, s: Statement) -> bool:
        if len(s) == 1:
            return True
        if len(s) == 2 and s[1].typ == Token.NUMBER:
            return True
        if (len(s) == 3
                and s[1].typ in (Token.NUMBER, Token.DASH)
                and s[2].typ == Token.NUMBER):
            return True
        return self._usage()

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        mrg = pt2iu(_SHRINK_DEFAULT)
        if len(s) >= 2 and s[1].typ == Token.NUMBER:
            mrg = pt2iu(s[1].num)

        has_ratio = len(s) >= 3
        ratio     = s[2].num if has_ratio else 1.0

        actual  = f.full_bbox()   # paper coords of panel content at present
        desired = f.extent()      # desired paper coords of panel

        dleft   = actual.left()   - desired.left()    # +ve means content is inside
        dright  = desired.right() - actual.right()
        dtop    = actual.top()    - desired.top()
        dbottom = desired.bottom()- actual.bottom()

        x0 = QPointF(f.x_axis().minp())   # paper coords of data range
        x1 = QPointF(f.x_axis().maxp())
        y0 = QPointF(f.y_axis().minp())
        y1 = QPointF(f.y_axis().maxp())

        old_dx = x1.x() - x0.x()   # paper width of data area
        old_dy = y1.y() - y0.y()   # paper height of data area

        # Push axes inward wherever content overflows the margin
        if dleft < mrg:
            x0 += QPointF(2 * mrg - dleft, 0)
        if dright < mrg:
            x1 -= QPointF(2 * mrg - dright, 0)
        if dtop < mrg:
            y1 += QPointF(0, 2 * mrg - dtop)
        if dbottom < mrg:
            y0 -= QPointF(0, 2 * mrg - dbottom)

        new_dx = x1.x() - x0.x()
        new_dy = y1.y() - y0.y()

        # Fudge failure: axes have flipped sign (collapsed past zero)
        if new_dx * old_dx < 0 or new_dy * old_dy < 0:
            Error() << (f"Fudge failure ({old_dx},{old_dy}) "
                        f"-> ({new_dx},{new_dy})")
            f.mark_fudge_failure()
            return

        # Optionally enforce an aspect ratio
        if has_ratio:
            dpx  = math.fabs(new_dx)
            dpy  = math.fabs(new_dy)
            xrat = math.fabs(dpx / (f.x_axis().max() - f.x_axis().min()))
            yrat = math.fabs(dpy / (f.y_axis().max() - f.y_axis().min()))
            myrat = yrat / xrat
            if myrat > ratio:           # too tall — shrink height
                adj = 0.5 * dpy * (1 - ratio / myrat)
                y1 += QPointF(0,  adj)
                y0 -= QPointF(0,  adj)
            elif myrat < ratio:         # too wide — shrink width
                adj = 0.5 * dpx * (1 - myrat / ratio)
                x1 -= QPointF(adj, 0)
                x0 += QPointF(adj, 0)

        # Only mark fudged if axes actually moved by more than the margin
        def manhattan(p: QPointF) -> float:
            return math.fabs(p.x()) + math.fabs(p.y())

        if (manhattan(x0 - f.x_axis().minp()) > mrg
                or manhattan(x1 - f.x_axis().maxp()) > mrg
                or manhattan(y0 - f.y_axis().minp()) > mrg
                or manhattan(y1 - f.y_axis().maxp()) > mrg):
            f.x_axis().set_placement(x0, x1)
            f.y_axis().set_placement(y0, y1)
            f.mark_fudged()
